src/analog.rs

```diff
@@ -0,0 +1,50 @@
+//! Analog-digital conversion traits
+
+use nb;
+
+/// This trait denotes an object, e.g. a GPIO pin, that is configured for use as an ADC channel.
+/// Objects of this type are passed to ADC-mode-specific methods.
+pub trait AdcChannel {
+    /// Channel ID type
+    ///
+    /// e.g. u8 for ADC channels 1-n, or a (partid, pinid) tuple for something more complex.
+    type ID;
+
+    /// ID of the channel associated with this pin
+    const CHANNEL: Self::ID;
+}
+
+/// ADCs that sample on single channels per request, and do so at the time of the request.
+pub trait SingleMode<Word, Pin: AdcChannel> {
+    /// Error type returned by ADC methods
+    type Error;
+
+    /// Take an analog reading from the specified pin
+    fn read_channel(&mut self, pin: &mut Pin) -> nb::Result<Word, Self::Error>;
+}
+
+/// ADCs that sample continuously without requiring requests
+pub trait Continuous<Word, Pin: AdcChannel> {
+    /// Error type returned by ADC methods
+    type Error;
+
+    /// Take ownership of an ADC channel/pin
+    fn add_channel(&mut self, pin: Pin) -> nb::Result<(), Self::Error>;
+
+    /// Release the ADC channel/pin
+    fn release_channel(&mut self) -> nb::Result<Pin, Self::Error>;
```

Comment:
Presumably you can call `add_channel` multiple times. At least the trait won't prevent that. What will `release_channel` return then?

Reply:
D'oh. This, too, should take an `AdcChannel::CHANNEL` argument to release the owned pin.

Reply:
Sorry, mistake here. `Continuous::release_channel` should indeed take a `Pin::ID` argument.